analyzers/src/SonarAnalyzer.Common/Rules/UseDateTimeInsteadOfDateTimeOffsetBase.cs

```diff
@@ -0,0 +1,72 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules;
+
+public abstract class UseDateTimeInsteadOfDateTimeOffsetBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+    where TSyntaxKind : struct
+{
+    private const string DiagnosticId = "S6566";
+
+    protected override string MessageFormat => "Prefer using \"DateTimeOffset\" struct instead of \"DateTime\"";
+
+    private static readonly ImmutableArray<string> TargetMemberAccess = ImmutableArray.Create(
+            "MaxValue",
+            "MinValue",
+            "UnixEpoch",
+            "Now",
+            "Today",
+            "UtcNow");
+
+    protected abstract bool IsNamedDateTime(SyntaxNode node);
+
+    protected UseDateTimeInsteadOfDateTimeOffsetBase() : base(DiagnosticId) { }
+
+    protected sealed override void Initialize(SonarAnalysisContext context)
+    {
+        context.RegisterNodeAction(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    if (IsDateTimeType(c.Node, c.SemanticModel))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                    }
+                },
+                Language.SyntaxKind.ObjectCreationExpressions);
+
+        context.RegisterNodeAction(
+            Language.GeneratedCodeRecognizer,
+            c =>
+            {
+                if (Language.Syntax.NodeExpression(c.Node) is var expression
+                    && IsNamedDateTime(expression)
+                    && TargetMemberAccess.Contains(Language.GetName(c.Node))
+                    && IsDateTimeType(expression, c.SemanticModel))
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                }
+            },
+            Language.SyntaxKind.SimpleMemberAccessExpression);
```

Comment:
The two registration can be merged if you
1. Register for `Language.SyntaxKind.IdentifierName` instead.
2. Check if the name is `DateTime` or `System.DateTime` (or `Date` in VB).
3. If it is then do a symbolic check to see if it's actually `System.DateTime`.
4. Mark certain edge cases (e.g. `System.DateTime` is aliased with the using directive) as FN.

This will help in covering some of the additional scenarios: method parameters, field and property declarations.

Reply:
This way you will miss all the implicit object creations if I'm not mistaken:
`DateTime date = new(1);`

Reply:
Yes you are right indeed. Still I'm not sure because we don't want to raise on every System.DateTime identifier:
```cs
DateTime.IsLeapYear(1);
        DateTime.Parse("06/01/1993");
        DateTime.ParseExact("06/01/1993", "dd/MM/yyyy", null);
        DateTime.SpecifyKind(date, DateTimeKind.Local);
        var typeName = nameof(DateTime); // Compliant
```

So you need to do the same checks in reverse to see whether they are indeed the noncompliant creation/access expression. IDK, I think it's fine to register directly for both and have a simple logic that checks the children instead of the parents/context.

Reply:
There are a lot of Identifiers in a source code, so in my opinion, it is better to have the two separate registrations. It should perform better.